src/ast/semantic_analyser.cpp

```diff
@@ -879,20 +879,47 @@ void SemanticAnalyser::visit(Call &call)
   }
   else if (call.func == "time") {
     check_assignment(call, false, false, false);
-    if (check_varargs(call, 0, 1)) {
-      if (is_final_pass()) {
-        if (call.vargs && call.vargs->size() > 0) {
+    if (check_varargs(call, 0, 2))
+    {
+      if (is_final_pass())
+      {
+        if (call.vargs == nullptr || call.vargs->size() == 0)
+        {
+          std::string fmt_default = "%H:%M:%S\n";
+          bpftrace_.time_args_.push_back(fmt_default.c_str());
+          warning("Using `time()` to print time directly is now deprecated.",
+                  call.loc);
+        }
+        else if (call.vargs->size() == 1)
+        {
           if (check_arg(call, Type::string, 0, true)) {
             auto &fmt_arg = *call.vargs->at(0);
             String &fmt = static_cast<String&>(fmt_arg);
             bpftrace_.time_args_.push_back(fmt.str);
+            warning("Using `time()` to print time directly is now deprecated.",
+                    call.loc);
+          }
+        }
+        else
+        {
+          if (check_arg(call, Type::string, 0, true) &&
+              check_arg(call, Type::integer, 1, false))
+          {
+            auto &fmt_arg = *call.vargs->at(0);
+            String &fmt = static_cast<String&>(fmt_arg);
+            bpftrace_.timestamp_args_.push_back(fmt.str);
           }
-        } else {
-          std::string fmt_default = "%H:%M:%S\n";
-          bpftrace_.time_args_.push_back(fmt_default.c_str());
         }
       }
     }
+    if (call.vargs == nullptr || call.vargs->size() < 2)
+    {
+      call.type = CreateNone();
+    }
+    else
+    {
+      call.type = CreateTimestamp();
+    }
```

Comment:
Can't you just have the `call.type = CreateTimestamp()` right after `bpftrace_.timestamp_args_.push_back(fmt.str)` and omit this entirely?

Reply:
I was thinking that doing so (putting `call.type = CreateTimestamp()` inside the `if (is_final_pass())` block) would result in `call.type` having 'Type::none' in previous passes. I assumed it would be prudent to let it have the correct type for all passes.

Reply:
Ah right. Can you move the `is_final_pass()` into the inner arms of if-branches, then? It'd be nice to keep all the type-dependent logic close to each other.